File: VB/T306403/CustomColorEdit/CustomPopupColorBuilder.vb
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraTab
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Namespace T306403.CustomColorEdit
    Public Class CustomPopupColorBuilder
        Inherits PopupColorBuilder

        Public Sub New(ByVal edit As IPopupColorEdit)
            MyBase.New(edit)

        End Sub
        Protected Overrides Sub SetTabPageProperties(ByVal pageIndex As Integer, ByVal shadowForm As PopupBaseForm)
            If pageIndex = 3 Then
                Dim tabPage As XtraTabPage = Me.TabControl.TabPages(pageIndex)
                Dim colorBox As ColorListBox = CreateColorListBox()
                colorBox.HighlightedItemStyle = Properties.HighlightedItemStyle
                Dim colors(((TryCast(Me.Properties, RepositoryItemCustomColorEdit)).ColorSource.Count) - 1) As Object
                For i As Integer = 0 To colors.Length - 1
                    colors(i) = (TryCast(Me.Properties, RepositoryItemCustomColorEdit)).ColorSource(i)
                Next i
                colorBox.Items.AddRange(colors)
                AddHandler colorBox.EnterColor, AddressOf OnEnterColor
                AddHandler colorBox.SelectedIndexChanged, AddressOf OnSelectedIndexChanged
                colorBox.Size = GetBestListBoxSize(colorBox)
                tabPage.PageVisible = (TryCast(Me.Properties, RepositoryItemCustomColorEdit)).ShowMyCustomColors
                colorBox.Dock = DockStyle.Fill
                colorBox.BorderStyle = BorderStyles.NoBorder
                If Owner.LookAndFeel IsNot Nothing Then
                    colorBox.LookAndFeel.Assign(Owner.LookAndFeel)
                End If
                tabPage.Controls.Add(colorBox)
                Return
            End If
            MyBase.SetTabPageProperties(pageIndex, shadowForm)
        End Sub
    End Class
End Namespace

File: CS/T306403/CustomColorEdit/CustomPopupColorBuilder.cs
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraTab;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace T306403.CustomColorEdit {
    public class CustomPopupColorBuilder : PopupColorBuilder {
        public CustomPopupColorBuilder(IPopupColorEdit edit)
            : base(edit) {

        }
        protected override void SetTabPageProperties(int pageIndex, PopupBaseForm shadowForm) {
            if(pageIndex == 3) {
                XtraTabPage tabPage = this.TabControl.TabPages[pageIndex];
                ColorListBox colorBox = CreateColorListBox();
                colorBox.HighlightedItemStyle = Properties.HighlightedItemStyle;
                object[] colors = new object[(this.Properties as RepositoryItemCustomColorEdit).ColorSource.Count];
                for(int i = 0; i < colors.Length; i++) {
                    colors[i] = (this.Properties as RepositoryItemCustomColorEdit).ColorSource[i];
                }
                colorBox.Items.AddRange(colors);
                colorBox.EnterColor += new EnterColorEventHandler(OnEnterColor);
                colorBox.SelectedIndexChanged += new EventHandler(OnSelectedIndexChanged);
                colorBox.Size = GetBestListBoxSize(colorBox);
                tabPage.PageVisible = (this.Properties as RepositoryItemCustomColorEdit).ShowMyCustomColors;
                colorBox.Dock = DockStyle.Fill;
                colorBox.BorderStyle = BorderStyles.NoBorder;
                if(Owner.LookAndFeel != null)
                    colorBox.LookAndFeel.Assign(Owner.LookAndFeel);
                tabPage.Controls.Add(colorBox);
                return;
            }
            base.SetTabPageProperties(pageIndex, shadowForm);
        }
    }
}
